File: services/ingestion/api/routes/ingest.py
"""Ingestion API routes."""

import logging
from uuid import UUID, uuid4

from api.dependencies import get_current_user, get_job_tracker
from api.schemas import (
    ActiveJobsResponse,
    CancelJobResponse,
    IngestRequest,
    IngestResponse,
    JobProgress,
    JobStatus,
    JobStatusResponse,
    ReembedRequest,
    ReembedResponse,
    SingleIngestRequest,
    SyncRequest,
    SyncResponse,
)
from fastapi import APIRouter, Depends, HTTPException
from tasks.ingest import batch_ingest, process_document
from tasks.models import JobStatus as TaskJobStatus
from tasks.reembed import reembed_collection

logger = logging.getLogger(__name__)

router = APIRouter()


@router.post(
    "",
    response_model=IngestResponse,
    status_code=202,
    summary="Start ingestion job",
    description="Trigger an async ingestion job for documents from the specified source.",
)
async def start_ingestion(
    request: IngestRequest,
    current_user: dict = Depends(get_current_user),
) -> IngestResponse:
    """
    Start a new ingestion job.

    The job runs asynchronously and returns immediately with a job ID
    that can be used to track progress.

    **Source Types:**
    - `filesystem`: Ingest from local files or S3
    - `database`: Ingest from PostgreSQL or MySQL
    - `web`: Crawl and ingest web pages
    - `api`: Ingest from REST API

    **Returns:**
    - `job_id`: UUID to track the job
    - `status`: Initial status (always "pending")
    """
    # Validate tenant access
    if request.acl.tenant_id != current_user.get("tenant_id"):
        raise HTTPException(
            status_code=403,
            detail="Cannot ingest documents for another tenant",
        )

    # Generate job ID
    job_id = uuid4()

    # Start async job
    task = batch_ingest.delay(
        job_id=str(job_id),
        source_type=request.source_type.value,
        source_config=request.source_config,
        processing_config=request.processing.model_dump(),
        acl_context=request.acl.model_dump(),
    )

    logger.info(
        f"Started ingestion job {job_id}",
        extra={
            "job_id": str(job_id),
            "source_type": request.source_type.value,
            "tenant_id": request.acl.tenant_id,
        },
    )

    return IngestResponse(
        job_id=UUID(task.id),
        status="pending",
        message="Ingestion job started",
    )


@router.post(
    "/single",
    response_model=IngestResponse,
    status_code=202,
    summary="Ingest single document",
    description="Ingest a single document by source ID.",
)
async def ingest_single_document(
    request: SingleIngestRequest,
    current_user: dict = Depends(get_current_user),
) -> IngestResponse:
    """
    Ingest a single document.

    Useful for incremental ingestion or re-processing specific documents.
    """
    # Validate tenant access
    if request.acl.tenant_id != current_user.get("tenant_id"):
        raise HTTPException(
            status_code=403,
            detail="Cannot ingest documents for another tenant",
        )

    task = process_document.delay(
        document_source_id=request.source_id,
        source_type=request.source_type,
        source_config=request.source_config,
        processing_config=request.processing.model_dump(),
        acl_context=request.acl.model_dump(),
    )

    return IngestResponse(
        job_id=UUID(task.id),
        status="pending",
        message="Document ingestion started",
    )


@router.get(
    "/{job_id}",
    response_model=JobStatusResponse,
    summary="Get job status",
    description="Get the current status and progress of an ingestion job.",
)
async def get_job_status(
    job_id: UUID,
    job_tracker=Depends(get_job_tracker),
) -> JobStatusResponse:
    """
    Get the status of an ingestion job.

    **Status Values:**
    - `pending`: Job is queued
    - `started`: Job has started processing
    - `progress`: Job is in progress (check `progress` field)
    - `success`: Job completed successfully
    - `failure`: Job failed (check `error_message`)
    - `revoked`: Job was cancelled
    """
    status = await job_tracker.get_job_status(str(job_id))

    if status is None:
        raise HTTPException(
            status_code=404,
            detail=f"Job {job_id} not found",
        )

    # Map task status to API status
    status_map = {
        TaskJobStatus.PENDING: JobStatus.PENDING,
        TaskJobStatus.STARTED: JobStatus.STARTED,
        TaskJobStatus.PROGRESS: JobStatus.PROGRESS,
        TaskJobStatus.SUCCESS: JobStatus.SUCCESS,
        TaskJobStatus.FAILURE: JobStatus.FAILURE,
        TaskJobStatus.REVOKED: JobStatus.REVOKED,
    }
    api_status = status_map.get(status.status, JobStatus.PENDING)

    # Calculate percentage if in progress
    progress = None
    if status.progress and status.progress.total > 0:
        progress = JobProgress(
            current=status.progress.current,
            total=status.progress.total,
            stage=status.progress.stage,
            percentage=round(status.progress.current / status.progress.total * 100, 1),
        )

    return JobStatusResponse(
        job_id=job_id,
        status=api_status,
        progress=progress,
        documents_processed=status.documents_processed,
        chunks_created=status.chunks_created,
        started_at=status.started_at,
        completed_at=status.completed_at,
        duration_seconds=status.duration_seconds,
        error_message=status.error_message,
        errors=status.errors,
    )


@router.delete(
    "/{job_id}",
    response_model=CancelJobResponse,
    summary="Cancel job",
    description="Cancel a running or pending ingestion job.",
)
async def cancel_job(
    job_id: UUID,
    current_user: dict = Depends(get_current_user),
    job_tracker=Depends(get_job_tracker),
) -> CancelJobResponse:
    """
    Cancel an ingestion job.

    Only pending and running jobs can be cancelled.
    """
    status = await job_tracker.get_job_status(str(job_id))

    if status is None:
        raise HTTPException(
            status_code=404,
            detail=f"Job {job_id} not found",
        )

    if status.status in [TaskJobStatus.SUCCESS, TaskJobStatus.FAILURE]:
        raise HTTPException(
            status_code=400,
            detail="Cannot cancel a completed job",
        )

    success = await job_tracker.cancel_job(str(job_id))

    return CancelJobResponse(job_id=job_id, cancelled=success)


@router.get(
    "",
    response_model=ActiveJobsResponse,
    summary="List active jobs",
    description="List all active ingestion jobs.",
)
async def list_active_jobs(
    job_tracker=Depends(get_job_tracker),
    current_user: dict = Depends(get_current_user),
) -> ActiveJobsResponse:
    """
    List all currently active ingestion jobs.

    Returns job IDs that are pending, started, or in progress.
    """
    job_ids = await job_tracker.list_active_jobs()

    # Get status for each job
    jobs = []
    for job_id in job_ids:
        status = await job_tracker.get_job_status(job_id)
        if status:
            # Map task status to API status
            status_map = {
                TaskJobStatus.PENDING: JobStatus.PENDING,
                TaskJobStatus.STARTED: JobStatus.STARTED,
                TaskJobStatus.PROGRESS: JobStatus.PROGRESS,
                TaskJobStatus.SUCCESS: JobStatus.SUCCESS,
                TaskJobStatus.FAILURE: JobStatus.FAILURE,
                TaskJobStatus.REVOKED: JobStatus.REVOKED,
            }
            api_status = status_map.get(status.status, JobStatus.PENDING)

            jobs.append(
                JobStatusResponse(
                    job_id=UUID(job_id),
                    status=api_status,
                    documents_processed=status.documents_processed,
                    chunks_created=status.chunks_created,
                    started_at=status.started_at,
                    completed_at=status.completed_at,
                    duration_seconds=status.duration_seconds,
                    error_message=status.error_message,
                    errors=status.errors,
                ),
            )

    return ActiveJobsResponse(jobs=jobs, total=len(jobs))


@router.post(
    "/sync",
    response_model=SyncResponse,
    status_code=202,
    summary="Trigger incremental sync",
    description="Start an incremental sync job for a source with updated_since filter.",
)
async def start_sync(
    request: SyncRequest,
    current_user: dict = Depends(get_current_user),
) -> SyncResponse:
    """
    Start an incremental sync for a source.

    This endpoint triggers a sync job that only processes documents
    updated since the specified timestamp. Useful for keeping the
    index up-to-date with source changes.

    **Source Types:**
    - `DATABASE`: Sync from database table with updated_since filter
    - `FILESYSTEM`: Sync modified files from path
    - `WEB`: Re-crawl pages with changes
    - `API`: Sync new/modified records from API
    """
    # Validate tenant access
    if request.tenant_id != current_user.get("tenant_id"):
        raise HTTPException(
            status_code=403,
            detail="Cannot sync documents for another tenant",
        )

    # Generate job ID
    job_id = uuid4()

    # Build source config with updated_since filter
    source_config = request.source_config.model_dump(exclude_none=True)

    # Start async sync job (uses existing batch_ingest with sync mode)
    task = batch_ingest.delay(
        job_id=str(job_id),
        source_type=request.source_type.value,
        source_config=source_config,
        processing_config={"mode": "incremental_sync"},
        acl_context={"tenant_id": request.tenant_id},
    )

    logger.info(
        f"Started sync job {job_id}",
        extra={
            "job_id": str(job_id),
            "source_type": request.source_type.value,
            "tenant_id": request.tenant_id,
            "updated_since": str(request.source_config.updated_since),
        },
    )

    return SyncResponse(
        job_id=UUID(task.id),
        status="queued",
        message="Incremental sync job started",
    )


@router.post(
    "/reembed",
    response_model=ReembedResponse,
    status_code=202,
    summary="Start re-embedding job",
    description="Start a re-embedding job with a new embedding model.",
)
async def start_reembed(
    request: ReembedRequest,
    current_user: dict = Depends(get_current_user),
) -> ReembedResponse:
    """
    Start a re-embedding job with a new model.

    This endpoint triggers a job to re-embed documents with a new
    embedding model. Useful for model migrations when upgrading
    to newer, better performing models.

    **Use Cases:**
    - Upgrading from bge-base to bge-large
    - Switching from English-only to multilingual model
    - Re-embedding after fixing preprocessing issues
    """
    # Validate tenant access if scope specifies tenant
    scope_tenant = request.target_scope.tenant_id
    if scope_tenant and scope_tenant != current_user.get("tenant_id"):
        raise HTTPException(
            status_code=403,
            detail="Cannot re-embed documents for another tenant",
        )

    # If no tenant specified in scope, use current user's tenant
    tenant_id = scope_tenant or current_user.get("tenant_id")

    # Generate IDs
    job_id = uuid4()
    embedding_job_id = uuid4()

    # Start async reembed job
    task = reembed_collection.delay(
        collection_name="rag_chunks",  # Default collection
        new_model=request.embedding_model,
        batch_size=request.batch_size,
        tenant_id=tenant_id,
    )

    logger.info(
        f"Started re-embedding job {embedding_job_id}",
        extra={
            "job_id": str(job_id),
            "embedding_job_id": str(embedding_job_id),
            "embedding_model": request.embedding_model,
            "tenant_id": tenant_id,
        },
    )

    return ReembedResponse(
        job_id=UUID(task.id),
        embedding_job_id=embedding_job_id,
        status="pending",
        message=f"Re-embedding job started with model {request.embedding_model}",
    )

File: crates/rag-ingestion/src/api/routes/ingest.rs
//! Ingestion job route handlers.

use std::sync::Arc;

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use chrono::Utc;
use serde_json::json;
use uuid::Uuid;

use crate::api::error::{ApiError, ApiResult};
#[cfg(test)]
use crate::api::jobs::JobTracker;
use crate::api::state::AppState;
use crate::api::types::{
    ActiveJobsResponse, CancelJobResponse, IngestRequest, IngestResponse, JobProgress, JobStatus,
    JobStatusResponse, ReembedRequest, ReembedResponse, SingleIngestRequest, SyncRequest,
    SyncResponse,
};
use crate::worker::{Job, JobPriority};

/// POST /api/v1/ingest - Start a batch ingestion job.
pub async fn start_ingestion(
    State(state): State<Arc<AppState>>,
    Json(request): Json<IngestRequest>,
) -> ApiResult<(StatusCode, Json<IngestResponse>)> {
    // Create job in tracker
    let job_id = state.job_tracker.create_job(request.acl.tenant_id.clone());

    tracing::info!(
        job_id = %job_id,
        source_type = ?request.source_type,
        tenant_id = %request.acl.tenant_id,
        "Started ingestion job"
    );

    // Enqueue job to Redis for async processing
    if let Some(job_queue) = &state.job_queue {
        let worker_job = Job::new(
            "ingest_batch",
            &request.acl.tenant_id,
            json!({
                "tracker_job_id": job_id.to_string(),
                "source_type": request.source_type,
                "source_config": request.source_config,
                "processing": {
                    "chunking_strategy": request.processing.chunking_strategy,
                    "chunk_size": request.processing.chunk_size,
                    "chunk_overlap": request.processing.chunk_overlap,
                    "enable_pii_detection": request.processing.enable_pii_detection,
                    "custom_metadata": request.processing.custom_metadata
                },
                "acl": {
                    "tenant_id": request.acl.tenant_id,
                    "visibility": request.acl.visibility,
                    "allowed_groups": request.acl.allowed_groups,
                    "allowed_users": request.acl.allowed_users
                }
            }),
        )
        .with_priority(JobPriority::Normal)
        .with_metadata("tracker_job_id", json!(job_id.to_string()));

        if let Err(e) = job_queue.enqueue(&worker_job).await {
            tracing::error!(error = %e, job_id = %job_id, "Failed to enqueue job to Redis");
            state
                .job_tracker
                .fail_job(&job_id, format!("Failed to enqueue: {e}"));
            return Err(ApiError::internal(format!("Failed to queue job: {e}")));
        }

        tracing::info!(
            job_id = %job_id,
            worker_job_id = %worker_job.id,
            "Job enqueued to Redis"
        );
    } else {
        tracing::warn!(job_id = %job_id, "No job queue configured - job will remain pending");
    }

    Ok((
        StatusCode::ACCEPTED,
        Json(IngestResponse {
            job_id,
            status: "pending".into(),
            message: "Ingestion job started".into(),
            created_at: Utc::now(),
        }),
    ))
}

/// POST /api/v1/ingest/single - Ingest a single document.
pub async fn ingest_single_document(
    State(state): State<Arc<AppState>>,
    Json(request): Json<SingleIngestRequest>,
) -> ApiResult<(StatusCode, Json<IngestResponse>)> {
    let job_id = state.job_tracker.create_job(request.acl.tenant_id.clone());

    tracing::info!(
        job_id = %job_id,
        source_id = %request.source_id,
        tenant_id = %request.acl.tenant_id,
        "Started single document ingestion"
    );

    // Enqueue job to Redis for async processing
    if let Some(job_queue) = &state.job_queue {
        let worker_job = Job::new(
            "ingest_single",
            &request.acl.tenant_id,
            json!({
                "tracker_job_id": job_id.to_string(),
                "source_type": request.source_type,
                "source_id": request.source_id,
                "source_config": request.source_config,
                "processing": {
                    "chunking_strategy": request.processing.chunking_strategy,
                    "chunk_size": request.processing.chunk_size,
                    "chunk_overlap": request.processing.chunk_overlap,
                    "enable_pii_detection": request.processing.enable_pii_detection,
                    "custom_metadata": request.processing.custom_metadata
                },
                "acl": {
                    "tenant_id": request.acl.tenant_id,
                    "visibility": request.acl.visibility,
                    "allowed_groups": request.acl.allowed_groups,
                    "allowed_users": request.acl.allowed_users
                }
            }),
        )
        .with_priority(JobPriority::Normal)
        .with_metadata("tracker_job_id", json!(job_id.to_string()));

        if let Err(e) = job_queue.enqueue(&worker_job).await {
            tracing::error!(error = %e, job_id = %job_id, "Failed to enqueue job to Redis");
            state
                .job_tracker
                .fail_job(&job_id, format!("Failed to enqueue: {e}"));
            return Err(ApiError::internal(format!("Failed to queue job: {e}")));
        }

        tracing::info!(
            job_id = %job_id,
            worker_job_id = %worker_job.id,
            "Job enqueued to Redis"
        );
    } else {
        tracing::warn!(job_id = %job_id, "No job queue configured - job will remain pending");
    }

    Ok((
        StatusCode::ACCEPTED,
        Json(IngestResponse {
            job_id,
            status: "pending".into(),
            message: "Document ingestion started".into(),
            created_at: Utc::now(),
        }),
    ))
}

/// GET /`api/v1/ingest/{job_id`} - Get job status.
pub async fn get_job_status(
    State(state): State<Arc<AppState>>,
    Path(job_id): Path<Uuid>,
) -> ApiResult<Json<JobStatusResponse>> {
    let job_state = state
        .job_tracker
        .get_job(&job_id)
        .ok_or_else(|| ApiError::not_found(format!("Job {job_id} not found")))?;

    let progress = if job_state.progress.total > 0 {
        Some(JobProgress {
            current: job_state.progress.current,
            total: job_state.progress.total,
            stage: job_state.progress.stage.clone(),
            percentage: job_state.progress.percentage,
        })
    } else {
        None
    };

    Ok(Json(JobStatusResponse {
        job_id,
        status: job_state.status,
        progress,
        documents_processed: job_state.documents_processed,
        chunks_created: job_state.chunks_created,
        started_at: job_state.started_at,
        completed_at: job_state.completed_at,
        duration_seconds: job_state.duration_seconds(),
        error_message: job_state.error_message,
        errors: job_state.errors,
    }))
}

/// DELETE /`api/v1/ingest/{job_id`} - Cancel a job.
pub async fn cancel_job(
    State(state): State<Arc<AppState>>,
    Path(job_id): Path<Uuid>,
) -> ApiResult<Json<CancelJobResponse>> {
    // Check if job exists
    let job_state = state
        .job_tracker
        .get_job(&job_id)
        .ok_or_else(|| ApiError::not_found(format!("Job {job_id} not found")))?;

    // Check if job can be cancelled
    if matches!(job_state.status, JobStatus::Success | JobStatus::Failure) {
        return Err(ApiError::bad_request("Cannot cancel a completed job"));
    }

    let cancelled = state.job_tracker.cancel_job(&job_id);

    Ok(Json(CancelJobResponse { job_id, cancelled }))
}

/// GET /api/v1/ingest - List active jobs.
pub async fn list_active_jobs(
    State(state): State<Arc<AppState>>,
) -> ApiResult<Json<ActiveJobsResponse>> {
    let job_ids = state.job_tracker.list_active_jobs();

    let jobs: Vec<JobStatusResponse> = job_ids
        .iter()
        .filter_map(|job_id| {
            state.job_tracker.get_job(job_id).map(|job_state| {
                let progress = if job_state.progress.total > 0 {
                    Some(job_state.progress.clone())
                } else {
                    None
                };

                JobStatusResponse {
                    job_id: *job_id,
                    status: job_state.status,
                    progress,
                    documents_processed: job_state.documents_processed,
                    chunks_created: job_state.chunks_created,
                    started_at: job_state.started_at,
                    completed_at: job_state.completed_at,
                    duration_seconds: job_state.duration_seconds(),
                    error_message: job_state.error_message,
                    errors: job_state.errors,
                }
            })
        })
        .collect();

    let total = jobs.len();
    Ok(Json(ActiveJobsResponse { jobs, total }))
}

/// POST /api/v1/ingest/sync - Start incremental sync.
///
/// Creates a tracked job and enqueues a `sync` task to Redis for background
/// processing. The worker will compare the source against the current index
/// and only re-ingest changed or new documents.
pub async fn start_sync(
    State(state): State<Arc<AppState>>,
    Json(request): Json<SyncRequest>,
) -> ApiResult<(StatusCode, Json<SyncResponse>)> {
    let job_id = state.job_tracker.create_job(request.tenant_id.clone());

    tracing::info!(
        job_id = %job_id,
        source_type = ?request.source_type,
        tenant_id = %request.tenant_id,
        "Started sync job"
    );

    // Enqueue sync job to Redis for async processing
    if let Some(job_queue) = &state.job_queue {
        let worker_job = Job::new(
            "sync",
            &request.tenant_id,
            json!({
                "tracker_job_id": job_id.to_string(),
                "tenant_id": request.tenant_id,
                "source_type": request.source_type,
                "source_config": request.source_config,
            }),
        )
        .with_priority(JobPriority::Normal)
        .with_metadata("tracker_job_id", json!(job_id.to_string()));

        if let Err(e) = job_queue.enqueue(&worker_job).await {
            tracing::error!(error = %e, job_id = %job_id, "Failed to enqueue sync job to Redis");
            state
                .job_tracker
                .fail_job(&job_id, format!("Failed to enqueue: {e}"));
            return Err(ApiError::internal(format!("Failed to queue sync job: {e}")));
        }

        tracing::info!(
            job_id = %job_id,
            worker_job_id = %worker_job.id,
            "Sync job enqueued to Redis"
        );
    } else {
        tracing::warn!(job_id = %job_id, "No job queue configured - sync job will remain pending");
    }

    Ok((
        StatusCode::ACCEPTED,
        Json(SyncResponse {
            job_id,
            status: "queued".into(),
            estimated_completion: None,
            message: "Incremental sync job started".into(),
        }),
    ))
}

/// POST /api/v1/ingest/reembed - Start re-embedding job.
///
/// Creates a tracked job and enqueues a `reembed` task to Redis for background
/// processing. The worker will re-generate embeddings for chunks matching the
/// target scope using the specified embedding model, then upsert the new
/// vectors into Qdrant.
pub async fn start_reembed(
    State(state): State<Arc<AppState>>,
    Json(request): Json<ReembedRequest>,
) -> ApiResult<(StatusCode, Json<ReembedResponse>)> {
    let tenant_id = request.target_scope.tenant_id.clone();
    if tenant_id.is_empty() {
        return Err(ApiError::bad_request(
            "target_scope.tenant_id must not be empty",
        ));
    }

    let job_id = state.job_tracker.create_job(tenant_id.clone());
    let embedding_job_id = Uuid::new_v4();

    tracing::info!(
        job_id = %job_id,
        embedding_job_id = %embedding_job_id,
        model = %request.embedding_model,
        tenant_id = %tenant_id,
        "Started re-embedding job"
    );

    // Enqueue reembed job to Redis for async processing
    if let Some(job_queue) = &state.job_queue {
        let worker_job = Job::new(
            "reembed",
            &tenant_id,
            json!({
                "tracker_job_id": job_id.to_string(),
                "embedding_job_id": embedding_job_id.to_string(),
                "embedding_model": request.embedding_model,
                "batch_size": request.batch_size,
                "target_scope": {
                    "tenant_id": request.target_scope.tenant_id,
                    "source_types": request.target_scope.source_types,
                    "document_ids": request.target_scope.document_ids,
                }
            }),
        )
        .with_priority(JobPriority::Normal)
        .with_metadata("tracker_job_id", json!(job_id.to_string()))
        .with_metadata("embedding_job_id", json!(embedding_job_id.to_string()));

        if let Err(e) = job_queue.enqueue(&worker_job).await {
            tracing::error!(error = %e, job_id = %job_id, "Failed to enqueue reembed job to Redis");
            state
                .job_tracker
                .fail_job(&job_id, format!("Failed to enqueue: {e}"));
            return Err(ApiError::internal(format!(
                "Failed to queue reembed job: {e}"
            )));
        }

        tracing::info!(
            job_id = %job_id,
            worker_job_id = %worker_job.id,
            embedding_job_id = %embedding_job_id,
            "Reembed job enqueued to Redis"
        );
    } else {
        tracing::warn!(
            job_id = %job_id,
            "No job queue configured - reembed job will remain pending"
        );
    }

    Ok((
        StatusCode::ACCEPTED,
        Json(ReembedResponse {
            job_id,
            embedding_job_id,
            status: "pending".into(),
            estimated_completion: None,
            message: format!(
                "Re-embedding job started with model {}",
                request.embedding_model
            ),
        }),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::api::types::{ACLContext, ProcessingOptions, SourceType, Visibility};

    fn test_state() -> Arc<AppState> {
        Arc::new(
            AppState::builder()
                .job_tracker(Arc::new(JobTracker::new()))
                .build()
                .unwrap(),
        )
    }

    #[tokio::test]
    async fn test_start_ingestion() {
        let state = test_state();
        let request = IngestRequest {
            source_type: SourceType::Filesystem,
            source_config: serde_json::json!({"path": "/data"}),
            processing: ProcessingOptions::default(),
            acl: ACLContext {
                tenant_id: "tenant-1".into(),
                visibility: Visibility::Private,
                allowed_groups: vec![],
                allowed_users: vec![],
            },
        };

        let result = start_ingestion(State(state.clone()), Json(request)).await;
        assert!(result.is_ok());

        let (status, response) = result.unwrap();
        assert_eq!(status, StatusCode::ACCEPTED);
        assert_eq!(response.status, "pending");

        // Verify job was created
        let job_state = state.job_tracker.get_job(&response.job_id);
        assert!(job_state.is_some());
    }

    #[tokio::test]
    async fn test_get_job_status() {
        let state = test_state();
        let job_id = state.job_tracker.create_job("tenant-1".into());

        let result = get_job_status(State(state), Path(job_id)).await;
        assert!(result.is_ok());

        let response = result.unwrap();
        assert_eq!(response.job_id, job_id);
        assert_eq!(response.status, JobStatus::Pending);
    }

    #[tokio::test]
    async fn test_get_job_status_not_found() {
        let state = test_state();
        let fake_id = Uuid::new_v4();

        let result = get_job_status(State(state), Path(fake_id)).await;
        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_cancel_job() {
        let state = test_state();
        let job_id = state.job_tracker.create_job("tenant-1".into());

        let result = cancel_job(State(state.clone()), Path(job_id)).await;
        assert!(result.is_ok());

        let response = result.unwrap();
        assert!(response.cancelled);

        // Verify job was cancelled
        let job_state = state.job_tracker.get_job(&job_id).unwrap();
        assert_eq!(job_state.status, JobStatus::Revoked);
    }

    #[tokio::test]
    async fn test_list_active_jobs() {
        let state = test_state();
        let _job1 = state.job_tracker.create_job("tenant-1".into());
        let _job2 = state.job_tracker.create_job("tenant-1".into());

        let result = list_active_jobs(State(state)).await;
        assert!(result.is_ok());

        let response = result.unwrap();
        assert_eq!(response.total, 2);
        assert_eq!(response.jobs.len(), 2);
    }
}
